plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/modelsourcewatcher/FileWatcher.java

```diff
@@ -0,0 +1,195 @@
+
+   @Inject
+   private ActionDispatcher actionDispatcher;
+
+   private final Map<String, List<FileWatchWorker>> workers = new HashMap<>();
+
+   @Inject
+   public FileWatcher(final ClientSessionManager sessionManager) {
+      sessionManager.addListener(this);
+   }
+
+   public FileWatcher(final ClientSessionManager sessionManager, final ActionDispatcher actionDispatcher) {
+      this(sessionManager);
+      this.actionDispatcher = actionDispatcher;
+   }
+
+   @Override
+   public void sessionClosed(final String clientId, final GLSPClient client) {
+      disposeAllWorkers(clientId);
+   }
+
+   @Override
+   public void startWatching(final GModelState modelState) {
+      createWorkers(modelState).forEach(FileWatchWorker::start);
+   }
+
+   @Override
+   public void stopWatching(final GModelState modelState) {
+      disposeAllWorkers(modelState.getClientId());
+   }
+
+   @Override
+   public void pauseWatching(final GModelState modelState) {
+      getAllWorkers(modelState.getClientId()).forEach(FileWatchWorker::pauseNotifications);
+   }
+
+   @Override
+   public void continueWatching(final GModelState modelState) {
+      getAllWorkers(modelState.getClientId()).forEach(FileWatchWorker::continueNotifications);
+   }
+
+   protected List<Path> getPaths(final GModelState modelState) {
+      final Optional<String> uriString = ClientOptions.getValue(modelState.getClientOptions(),
+         ClientOptions.SOURCE_URI);
+      URI sourceUri = URI.createFileURI(uriString.map(uri -> uri.replace(FILE, "")).orElseThrow());
+      sourceUri = URI.createURI(URI.decode(sourceUri.toString()));
+      final List<Path> paths = new ArrayList<>();
+      paths.add(new File(sourceUri.toFileString()).toPath());
+      return paths;
+   }
+
+   private List<FileWatchWorker> createWorkers(final GModelState modelState) {
+      stopAllWorkers(modelState.getClientId());
+      final List<FileWatchWorker> fileWorkers = createFileWatchWorkers(modelState);
+      workers.put(modelState.getClientId(), fileWorkers);
+      return fileWorkers;
+   }
+
+   private void disposeAllWorkers(final String clientId) {
+      stopAllWorkers(clientId);
+      workers.remove(clientId);
+   }
+
+   private void stopAllWorkers(final String clientId) {
+      getAllWorkers(clientId).forEach(FileWatchWorker::stopWorking);
+   }
+
+   protected Stream<FileWatchWorker> getAllWorkers(final String clientId) {
+      return Optional.ofNullable(workers.get(clientId)).stream().flatMap(Collection::stream);
+   }
+
+   private List<FileWatchWorker> createFileWatchWorkers(final GModelState modelState) {
+      return getPaths(modelState).stream().map(path -> new FileWatchWorker(modelState.getClientId(), path))
+         .collect(Collectors.toList());
+   }
+
+   protected void notifyClient(final String clientId, final Path filePath) {
+      actionDispatcher.dispatch(clientId, new ModelSourceChangedAction(filePath.toAbsolutePath().toString()));
```

Comment:
I noticed that sometimes we get two changes for the same file in quick succession. Maybe we should add a debounce here to avoid sending the same notification twice? Alternatively we can argue that the client must handle this.

Reply:
Yes, definitely! I added a debouncer.